File: ontrack-job/src/main/java/net/nemerosa/ontrack/job/orchestrator/JobOrchestrator.java
package net.nemerosa.ontrack.job.orchestrator;

import net.nemerosa.ontrack.job.*;

import java.util.Collection;
import java.util.HashSet;
import java.util.Set;
import java.util.stream.Collectors;

public class JobOrchestrator implements Job {

    private final JobScheduler jobScheduler;
    private final String name;
    private final Collection<JobOrchestratorSupplier> jobOrchestratorSuppliers;

    private final Set<JobKey> cache = new HashSet<>();

    public JobOrchestrator(JobScheduler jobScheduler, String name, Collection<JobOrchestratorSupplier> jobOrchestratorSuppliers) {
        this.jobScheduler = jobScheduler;
        this.name = name;
        this.jobOrchestratorSuppliers = jobOrchestratorSuppliers;
    }

    @Override
    public JobKey getKey() {
        return JobCategory.CORE.getType("orchestrator").withName("Orchestrator").getKey(name);
    }

    @Override
    public JobRun getTask() {
        return this::orchestrate;
    }

    public synchronized void orchestrate(JobRunListener runListener) {
        // Complete list of registrations
        Collection<JobRegistration> registrations = jobOrchestratorSuppliers.stream()
                .flatMap(JobOrchestratorSupplier::collectJobRegistrations)
                .collect(Collectors.toList());
        // List of keys
        Set<JobKey> keys = registrations.stream()
                .map(registration -> registration.getJob().getKey())
                .collect(Collectors.toSet());
        // Jobs to unschedule
        Set<JobKey> toRemove = new HashSet<>(cache);
        toRemove.removeAll(keys);
        toRemove.forEach(jobScheduler::unschedule);
        // Jobs to add
        Set<JobKey> toAdd = new HashSet<>(keys);
        toAdd.removeAll(cache);
        registrations.stream()
                .filter(jobRegistration -> toAdd.contains(jobRegistration.getJob().getKey()))
                .forEach(jobRegistration -> schedule(jobRegistration, runListener));
        // Resets the cache
        cache.clear();
        cache.addAll(keys);
    }

    private void schedule(JobRegistration jobRegistration, JobRunListener runListener) {
        runListener.message("Scheduling: %s", jobRegistration.getJob().getKey());
        jobScheduler.schedule(jobRegistration.getJob(), jobRegistration.getSchedule());
    }

    @Override
    public String getDescription() {
        return name;
    }

    @Override
    public boolean isDisabled() {
        return false;
    }

}

File: ontrack-job/src/main/java/net/nemerosa/ontrack/job/orchestrator/JobOrchestrator.kt
package net.nemerosa.ontrack.job.orchestrator

import net.nemerosa.ontrack.job.*
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.*
import kotlin.streams.toList
import org.springframework.transaction.PlatformTransactionManager
import org.springframework.transaction.support.TransactionTemplate

class JobOrchestrator(
    private val jobScheduler: JobScheduler,
    private val name: String,
    private val jobOrchestratorSuppliers: Collection<JobOrchestratorSupplier>,
    transactionManager: PlatformTransactionManager,
) : Job {

    private val cache = HashSet<JobKey>()
    private val transactionTemplate = TransactionTemplate(transactionManager)
    private val logger: Logger = LoggerFactory.getLogger(JobOrchestrator::class.java)

    override fun getKey(): JobKey {
        return JobCategory.CORE.getType("orchestrator").withName("Orchestrator").getKey(name)
    }

    override fun getTask(): JobRun {
        return JobRun { this.orchestrate(it) }
    }

    @Synchronized
    fun orchestrate(runListener: JobRunListener) {
        // Complete list of registrations
        val registrations = transactionTemplate.execute {
            jobOrchestratorSuppliers.flatMap {
                try {
                    it.jobRegistrations
                } catch (any: Exception) {
                    // Does not collect the jobs for this collector for now, but marks the exception
                    logger.error("Cannot collect jobs for ${it::class.java.name}", any)
                    emptyList()
                }
            }
        } ?: return
        // List of registration keys
        val keys = registrations.map { registration -> registration.job.key }
        // Jobs to unschedule
        val toRemove = HashSet(cache)
        toRemove.removeAll(keys)
        toRemove.forEach { jobScheduler.unschedule(it) }
        // Jobs to add / update
        val toRegister = HashSet(keys)
        toRegister.removeAll(toRemove)
        registrations
            .filter { jobRegistration -> toRegister.contains(jobRegistration.job.key) }
            .forEach { jobRegistration -> schedule(jobRegistration, runListener) }
        // Resets the cache
        cache.clear()
        cache.addAll(keys)
    }

    private fun schedule(jobRegistration: JobRegistration, runListener: JobRunListener) {
        runListener.message("Scheduling: %s", jobRegistration.job.key)
        jobScheduler.schedule(jobRegistration.job, jobRegistration.schedule)
    }

    override fun getDescription(): String {
        return name
    }

    override fun isDisabled(): Boolean {
        return false
    }

}
